### DISCUSSION TITLE
Possible Labels for Github

### DISCUSSION THREAD

**Discussion by bobgeorgethe3rd** [freesewing/freesewing#3952] [Support]
During the meeting 2023-04-29 we discussed the new labels on github such as changesRequested and awaitingFeedback and how it may be prudent to add some more labels and better establish to people what labels stand for. Here are some label ideas.

- docReviewRequired
Some who works on docs/part of doc teams need to approve before merging.

- informationRequested
When information is needed rather than changes. This would sort of replace awaitingFeedback and indicates to the PR submitter that there is action they need to take.

- changesMade
Tells the reviewers to look at the PR again as the changesRequested have been made.

- informationGiven
PR submitter has replied and given the informationRequested. This indicates the revievwer to take a look.

- doNotMerge
Quick flag to stop the PR from being merged if needed. e.g. discussion for the PR is happening elsewhere and needs to be reached before merger.

- Fixed in v3
To help close v2 bugs that are being fixed in v3 but will not be fixed in v2

What do we think? Love them? Hate them? Tweak them? Got other ones?

**Comment by BenJamesBen** [comment]
Just some thoughts:

(GitHub allows space characters in labels, so we aren't restricted to camelCaseWords.)

So, "Changes requested" means we are waiting for the PR submitter to make changes and update the PR?

What does "Awaiting feedback"? mean? Is Does it mean that it is waiting for reviewers to review the PR? If so, it might be better to change it to something like "Reviews requested". ("Awaiting Feedback" is ambiguous about who we're waiting for.)

Instead of using "Changes made" or "Information Given" labels, we could instead have the PR submitter remove the old labels andre-apply the "Reviews Requested" label for that situation.

Instead of "Do Not Merge", we could instead have "Ready to Merge", to indicate that all reviews are done and everything is ready to merge. Joost or whoever could then simply look for that label to know which PRs can be merged.

I was thinking of a generic "Reviews Requested" label would be simplier to use. However, if we want to split it into "Doc Reviews Requested" and "Code Reviews Requested", it might be useful for contributors who do only doc review or code review.

--

Also, for those v2 bugs that we want to close:
"Fixed in v3" or possibly more generically, so we can use it for v4, v5, etc.: "Fixed in next version".

**Reply by bobgeorgethe3rd** [reply]
I know we're not restricted by it but it might be an idea to restrict the labels intended for PRs in some way so they are established from other labels, and I was just using camelCase due to the existing labels. 

Yes "changesRequested" means someone has requested changes that need to be made by the submitter before merger.

As I have come to understand it "awaitingFeeback" means awaiting feedback from the submitter on a subject and acts as a skip over label when joost is look through the prs. I agreed that is kinda ambiguous and due to this I was actually using it incorrectly so that's why I am suggesting a different term like "infortmationRequested" so it is clearer it is the submitter who needs to make an action. It also could be "discussionNeeded" but that is possibly too open ended.

There is an arguement to have a catch all label for needing a review but the more specific "changesMade" & "informationGiven" is that it tells the reviewer immediately what actions have been taken and what they to check. Also "informationGiven" could lead into "changesNeeded". Yea the idea would be to take of the relevant labels and replace them with the new labels when changes/actions have been made.

"Ready to Merge" can mean different things to different people. For instance for me doc works requires a thorough proof-read before it is merged whereas for someone else this may just mean the doc work being PRed is ready to merge. This wouldn't stop PRs from going through that someone may wish to review but can't currently at the time. Essestainly "doNotMerge" or "doNotTouch" is hey please don't merge this yet as I have noticed something that needs addressing that I can't at this very moment in time or we are having a dicussion outside of the PR and it's just a yo Im looking into this please wait for me.

The reason why the docReviewRequired rather than generic was that someone who works with docs needs to approve this before it can be merged. This will stop documentation being merged without a proper proof-read and stuff like typos and formatting can be caught before it lands on the site. You can already request a review via githubs own system this marker is about who may need to approve rather than it needs to be reviewed.

Fixed in v3 should be good. I think it's better to indicate which version it was fixed in so if someones comes along later for whatever reason they don't have to try and figure out what version it came from/was fixed in.

**Reply by joostdecock** [reply]
> *it might be an idea to restrict the labels intended for PRs in some way so they are established from other labels*

I feel like it would make more sense to have a prefix that clarifies what status the PR is in. Eg something like:

- `PR Status: Needs Triage`
- `PR Status: Needs Review`
- `PR Status: Changes Requested`
- `PR Status: Info Requested`
- `PR Status: Needs approval (docs)`
- `PR Status: Needs approval (designs)`
- `PR Status: Needs approval (plugins)`
- `PR Status: Needs approval (core)`
- `PR Status: Needs approval (packages)`
- `PR Status: Needs approval (backend)`
- `PR Status: Needs approval (org/dev/lab)`
- `PR Status: Needs approval (other)`
- `PR Status: Ready to merge`

I think such a naming scheme has a number of benefits:

- It's a lot more self-explanatory
- It should be clear that a PR should only ever be in 1 status
- We can have each PR auto-tagged with the `PR Status: Needs Triage` badge (we can create a workflow for that)
- We could even auto-post a comment explaining the workflow and what statuses a PR should go through before it can be merged
- I divided the `Needs approval` labels based on what sort of thing it is (`docs`/`designs`/`plugins`/`core`/`packages`/`backend`/`org/dev/lab`/`other`) as typically these would be of interest to different people.

Perhaps we can have something similar for issues. Like:

- `Issue Status: Needs Triage`
- `Issue Status: Won't Fix`
- `Issue Status: Fixed in v3`
- `Issue Status: Unconfirmed`
- `Issue Status: Confirmed`
- `Issue Status: Fix suggested`
- `Issue Status: Feature Request Accepted`
- `Issue Status: Resolved`

I'm not saying this list is exhaustive, but in a similar way, issues should only ever be in 1 state.

**Reply by bobgeorgethe3rd** [reply]
I absolutely love this. My only gripe is there is a little lee-way with PR Status being more than one when it comes to approval as often PRS are making in changes in the different sections but I love this, it's so much clearer, it teaches anybody who comes in how it works and are expected workflow which is so cool.

**Reply by joostdecock** [reply]
Ok, let me take a stab at diagramming this (consider it a first draft):

```mermaid
flowchart TD

    core{{"Involves
core?"}}
    plugins{{"Involves
plugins?"}}
    designs{{"Involves
designs?"}}
    packages{{"Involves
(other)
packages?"}}
    backend{{"Involves
backend?"}}
    websites{{"Involves
websites?"}}
    docs{{"Involved
docs?"}}
    newPr([Pull Request Submitted]) --> needsTriage>PR Status: Needs Triage] 
    needsTriage --> notSpam{"Looks legit?"}  
    notSpam -- No --> close(["Close Pull Request"])
    notSpam --Yes--> ar["Assign/Request Reviewer(s)"]  
    -->  needsReview>PR Status: Needs Review]
    --> review{"Review passed?"} 
    review <-- lacksInfo --> infoRequested>PR Status: Info Requested ]
    review -- ok --> core
    review <-- needsChanges --> changesRequested>PR Status: Changes Requested ]
    
    core -- Yes --> needsApprovalCore>"PR Status: NeedsApproval (core)"]
    core -- No --> plugins
    needsApprovalCore --> coreOk{"Core OK?"}
    coreOk --Yes--> plugins
    coreOk --No--> needsReview
    
    plugins -- Yes --> needsApprovalPlugins>"PR Status: NeedsApproval (plugins)"]
    plugins -- No --> designs
    needsApprovalPlugins --> pluginsOk{"Plugins OK?"}
    pluginsOk --Yes--> designs
    pluginsOk --No--> needsReview
    
    designs -- Yes --> needsApprovalDesigns>"PR Status: NeedsApproval (designs)"]
    designs -- No --> packages
    needsApprovalDesigns --> designsOk{"Designs OK?"}
    designsOk --Yes--> packages
    designsOk --No--> needsReview

    packages -- Yes --> needsApprovalPackages>"PR Status: NeedsApproval (packages)"]
    packages -- No --> backend
    needsApprovalPackages --> packagesOk{"Packages OK?"}
    packagesOk --Yes--> backend
    packagesOk --No--> needsReview

    backend -- Yes --> needsApprovalBackend>"PR Status: NeedsApproval (backend)"]
    backend -- No --> websites
    needsApprovalBackend --> backendOk{"Backend OK?"}
    backendOk --Yes--> websites
    backendOk --No--> needsReview

    websites -- Yes --> needsApprovalWebsites>"PR Status: NeedsApproval (websites)"]
    websites -- No --> docs
    needsApprovalWebsites --> websitesOk{"Websites OK?"}
    websitesOk --Yes--> docs
    websitesOk --No--> needsReview

    docs -- Yes --> needsApprovalDocs>"PR Status: NeedsApproval (docs)"]
    docs -- No --> ready>"PR Status: Ready To Merge"]
    needsApprovalDocs --> docsOk{"Docs OK?"}
    docsOk --Yes--> ready --> merge(["Merge Pull Request"])
    docsOk --No--> needsReview
    classDef main fill:#d0bfff,stroke:#333,stroke-width:4px;
    classDef status fill:#63e6be,stroke:#333,stroke-width:2px;
    classDef check fill:#ffc078,stroke:#333,stroke-width:2px;
    classDef split fill:#ffe066,stroke:#333,stroke-width:2px;
  
    class newPr,merge,close main;
    class needsTriage,needsReview,infoRequested,changesRequested,needsApprovalCore,needsApprovalPlugins,needsApprovalDesigns,needsApprovalPackages,needsApprovalBackend,needsApprovalWebsites,needsApprovalDocs,ready status;
    class notSpam,review,coreOk,pluginsOk,packagesOk,designsOk,backendOk,websitesOk,docsOk check
    class core,plugins,packages,designs,backend,docs,website split
```
   
Not complicated at all 😂

**Comment by SeaZeeZee** [comment]
I think this is an amazing improvement on what we have now! I was working on reviewing some prs and was wondering, do we want a label for when things have been reviewed by someone without write access so the people who do have write access know they don't have to look at it if they don't have time/inclination?

**Comment by joostdecock** [comment]
So many ideas. So little time. Closing.
